### PR TITLE
[IAST] Call Site Instrumentation support

### PR DESCRIPTION
**Pull request by manuel-alvarez-alvarez**
# What Does This Do
This PR contains the basics to do Call Site Instrumentation in the tracer. It defines a new type of instrumenter `datadog.trace.agent.tooling.bytebuddy.csi.CallSiteInstrumenter` that is able to register instances of `datadog.trace.agent.tooling.csi.CallSiteAdvice` to perform the instrumentation via ASM.

# Motivation
Instrumentations related to IAST often affect core parts of the JDK (String, StringBuilder...) where using callee instrumentation (by default in ByteBuddy) is not good enough for performance. Call site instrumentation focuses on the calls to the instrumented methods enabling the use of inclusion/exclusion lists to fine tune where to instrument.

# Additional Notes
Following this PR comes [[IAST] Call Site Instrumentation plugin #3729](https://github.com/DataDog/dd-trace-java/pull/3729) where advices can be autogenerated by a gradle plugin. It also showcases an example on  how to use CSI in the tracer.

# Benchmark results
The main goal of this results are to validate that call site instrumentation is feasible in terms of startup time of an application:
| Benchmark  | Mode | Cnt | Score | Error | Units |
| ------------- | ------------- |------------- |------------- |------------- |------------- |
|CallSiteBenchmark.callSite   | ss  | 10 | 1572.961 | ± 36.189 | ms/op |
|CallSiteBenchmark.callee     | ss  | 10 | 1551.316 | ± 31.888 | ms/op |
|CallSiteBenchmark.none      |  ss  | 10 | 1337.549 | ± 52.882 | ms/op |

### CODE REVIEW

**Review by jandro996 — APPROVED** [review]

File: buildSrc/src/main/java/datadog/trace/plugin/csi/impl/AsmSpecificationBuilder.java [hunk 1]
```diff
@@ -0,0 +1,236 @@
+package datadog.trace.plugin.csi.impl;
```
Comment (anderruiz) [inline buildSrc/src/main/java/datadog/trace/plugin/csi/impl/AsmSpecificationBuilder.java:1]: This class does the parsing of the callbacks to search for @CallSite annotations to find the instrumentation spots

File: buildSrc/src/main/java/datadog/trace/plugin/csi/AdviceGenerator.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    @Override
+    public boolean isSuccess() {
+      return super.isSuccess() && advices.stream().allMatch(AdviceResult::isSuccess);
+    }
+
+    public List<AdviceResult> getAdvices() {
+      return advices;
+    }
```
Comment (ValentinZakharov) [inline buildSrc/src/main/java/datadog/trace/plugin/csi/AdviceGenerator.java:31]: Such uncertainty scares a bit, we are not sure how exactly caller will be using the list (just add? may be get? remove?)

File: buildSrc/src/main/java/datadog/trace/plugin/csi/impl/AsmSpecificationBuilder.java (hunk 1 above)
Comment (manuel-alvarez-alvarez) [reply buildSrc/src/main/java/datadog/trace/plugin/csi/impl/AsmSpecificationBuilder.java]: This PR has been divided in two, one for the instrumenter and another for the gradle plugin, this one now relates to just the instrumenter.

File: buildSrc/src/main/java/datadog/trace/plugin/csi/AdviceGenerator.java (hunk 2 above)
Comment (manuel-alvarez-alvarez) [reply buildSrc/src/main/java/datadog/trace/plugin/csi/AdviceGenerator.java]: This PR has been divided in two, one for the instrumenter and another for the gradle plugin, this one now relates to just the instrumenter.

**Review by smola — APPROVED** [review]

**Review by anderruiz — APPROVED** [review]

**Review by bantonsson — CHANGES_REQUESTED** [review]
Great PR!

There is only one thing that I think needs fixing in some way before merging.

I was a bit confused by the benchmarks, and it feels very counter intuitive that there should be any performance difference between `callSite` and `callee` in this case, so I had to poke at them a bit, and the response differs. The `callSite` one gets back `Hello! [Transformed]` and the `callee` one gets back `Hello! [Transformed] [Transformed]`. So we're not comparing the same things.

**Comment by manuel-alvarez-alvarez** [comment]
> Great PR!
> 
> There is only one thing that I think needs fixing in some way before merging.
> 
> I was a bit confused by the benchmarks, and it feels very counter intuitive that there should be any performance difference between `callSite` and `callee` in this case, so I had to poke at them a bit, and the response differs. The `callSite` one gets back `Hello! [Transformed]` and the `callee` one gets back `Hello! [Transformed] [Transformed]`. So we're not comparing the same things.

@bantonsson, first of all thanks for the feedback :smiley:, much appreciated!

My idea with the benchmark is to validate that doing call site instrumentation is feasible in terms of startup time as long as there's a proper exclusion list.

In this particular case I wanted to showcase the instrumentation of `javax.servlet.http.HttpServlet#getParameter` as it's something we need for IAST. It's true that when using callee, due to wrappers/facades/... you end up being executed a few times in the stack of a request (that's why we got the double `[Transformed]` response). I've changed to instrument only the real implementation of the interface (in this particular case it's the apache connector).

Let me know if I should do something else about it.

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/bytebuddy/csi/Advices.java [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+          final @Nonnull Advices advices,
+          final @Nonnull TypeDescription type,
+          final ClassLoader classLoader) {
+        return advices;
+      }
+    }
+
+    /**
+     * This class will try to parse the constant pool of the class in order to discover the required
+     * advices to parse it.
```
[the remark this answers is not in the document]
Comment (manuel-alvarez-alvarez) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/bytebuddy/csi/Advices.java]: Ops! my bad I don't know what I was ~~parsing~~ thinking when I wrote it, fixed!

**Review by bantonsson — APPROVED** [review]

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/bytebuddy/csi/CallSiteInstrumenter.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+import datadog.trace.agent.tooling.csi.CallSiteAdvice;
+import java.util.ServiceLoader;
+import javax.annotation.Nonnull;
+
+/**
+ * Instrumented using a {@link CallSiteTransformer} to perform the actual instrumentation. It will
+ * fetch the instance of the {@link CallSiteAdvice} from SPI according to the specified marker
+ * interface.
+ */
+public abstract class CallSiteInstrumenter extends Instrumenter.Default
```
[the remark this answers is not in the document]
Comment (manuel-alvarez-alvarez) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/bytebuddy/csi/CallSiteInstrumenter.java:13]: Exactly, couldn't have explained it better. Our idea is to provide a `CallSiteInstrumenter` that can be used by any other product if so desired.

**Review by DDJavierSantos — APPROVED** [review]

**Pull request merged**